<h1 align="center" style="background-color:beige;color:black">Data Cleaning</h1>

In [1]:
import pandas as pd
import numpy as np

# Data Accessing
In this step, the data is to be understood more deeply.Before implementing methods to clean it, you will definitely need to have a better idea about what the data is about.

# Types of Unclean Data

* **Dirty Data:** Dirty data, also known as low quality data has content issues.
<ol>
<li>Duplicated data.</li>
<li>Missing data.</li>
<li>Corrupt data.</li>
<li>Inaccurate data.</li>
</ol>

* **Messy Data:** Messy data, also known as untidy data. Untidy data has structural issues. Tidy data has the following properties.
<ol>
<li> Each variable forms a table.</li>
<li> Each observation forms a row.</li>
<li> Each observational unit forms a table.</li>
</ol>


In [2]:
df1=pd.read_csv("D:\Python\Datasets\patients.csv",index_col="patient_id")
df1.head()

assigned_sex given_name  surname                    address  \
patient_id                                                               
1                female        Zoe  Wellish       576 Brown Bear Drive   
2                female     Pamela     Hill  2370 University Hill Road   
3                  male        Jae   Debord      1493 Poling Farm Road   
4                  male       Liêm     Phan        2335 Webster Street   
5                  male        Tim  Neudorf       1428 Turkey Pen Lane   

                         city       state  zip_code        country  \
patient_id                                                           
1           Rancho California  California   92390.0  United States   
2                   Armstrong    Illinois   61812.0  United States   
3                        York    Nebraska   68467.0  United States   
4                  Woodbridge          NJ    7095.0  United States   
5                      Dothan          AL   36303.0  United States   

                                               contact  birthdate  weight  \
patient_id                                                                  
1                 951-719-9170ZoeWellish@superrito.com  7/10/1976   121.7   
2                PamelaSHill@cuvox.de+1 (217) 569-3204   4/3/1967   118.8   
3                     402-363-6804JaeMDebord@gustr.com  2/19/1980   177.8   
4           PhanBaLiem@jourrapide.com+1 (732) 636-8246  7/26/1951   220.9   
5                      334-515-7487TimNeudorf@cuvox.de  2/18/1928   192.3   

            height   bmi  
patient_id                
1               66  19.6  
2               66  19.2  
3               71  24.8  
4               70  31.7  
5               27  26.1

# Data Exploration

In [3]:
print(f'The number of features:{len(df1.columns)}')
print(df1.columns)
print(f'The shape of the dataset is:{df1.shape}\nThe size if the dataset is:{df1.size}')
print(df1.info())
print(df1.describe())
print(df1.info)
print(df1.isnull().sum())

The number of features:13
Index(['assigned_sex', 'given_name', 'surname', 'address', 'city', 'state',
       'zip_code', 'country', 'contact', 'birthdate', 'weight', 'height',
       'bmi'],
      dtype='object')
The shape of the dataset is:(503, 13)
The size if the dataset is:6539
<class 'pandas.core.frame.DataFrame'>
Index: 503 entries, 1 to 503
Data columns (total 13 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   assigned_sex  503 non-null    object 
 1   given_name    503 non-null    object 
 2   surname       503 non-null    object 
 3   address       491 non-null    object 
 4   city          491 non-null    object 
 5   state         491 non-null    object 
 6   zip_code      491 non-null    float64
 7   country       491 non-null    object 
 8   contact       491 non-null    object 
 9   birthdate     503 non-null    object 
 10  weight        503 non-null    float64
 11  height        503 non-null    int64  
 12  bmi       

<h4 align="center" style="background-color:beige;color:black">Number of missing values and percentage of missing values in each column</h4>

In [36]:
missing_df=df1.isnull().sum().to_frame().rename(columns={0:"Missing Values"})
missing_df["Percentage of missing values"]=(round((df1.isnull().sum()*100)/len(df1),2).astype(str)+ "%")
missing_df

Missing Values Percentage of missing values
assigned_sex               0                         0.0%
given_name                 0                         0.0%
surname                    0                         0.0%
address                   12                        2.39%
city                      12                        2.39%
state                     12                        2.39%
zip_code                  12                        2.39%
country                   12                        2.39%
contact                   12                        2.39%
birthdate                  0                         0.0%
weight                     0                         0.0%
height                     0                         0.0%
bmi                        0                         0.0%

In [5]:
df2=pd.read_csv("D:\\Python\\Datasets\\treatments.csv")
df2.head()

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
0   veronika    jindrová  41u - 48u          -         7.63       7.20   
1     elliot  richardson          -  40u - 45u         7.56       7.09   
2   yukitaka    takenaka          -  39u - 36u         7.68       7.25   
3       skye  gormanston  33u - 36u          -         7.97       7.62   
4     alissa      montez          -  33u - 29u         7.78       7.46   

   hba1c_change  
0           NaN  
1          0.97  
2           NaN  
3          0.35  
4          0.32

In [42]:
missingdf2=df2.isnull().sum().to_frame().rename(columns={0:"Missing Values"})
missingdf2["Percentage of missing values"]=(round((df2.isnull().sum()*100)/len(df2),2).astype(str)+ "%")
missingdf2


Missing Values Percentage of missing values
given_name                                 0                         0.0%
surname                                    0                         0.0%
auralin                                    0                         0.0%
novodra                                    0                         0.0%
hba1c_start                                0                         0.0%
hba1c_end                                  0                         0.0%
hba1c_change                             109                       38.93%
Percentage of Missing values             280                       100.0%

In [6]:
df3=pd.read_csv(r"D:\Python\Datasets\treatments_cut.csv")
df3.head()

given_name    surname    auralin    novodra  hba1c_start  hba1c_end  \
0      jožka  resanovič  22u - 30u          -         7.56       7.22   
1  inunnguaq   heilmann  57u - 67u          -         7.85       7.45   
2      alwin   svensson  36u - 39u          -         7.78       7.34   
3        thể      lương          -  61u - 64u         7.64       7.22   
4     amanda    ribeiro  36u - 44u          -         7.85       7.47   

   hba1c_change  
0          0.34  
1           NaN  
2           NaN  
3          0.92  
4          0.38

In [53]:
missingdf3=df3.isnull().sum().to_frame().rename(columns={0:"Missing Values"})
missingdf3["Percentage of missing values"]=(round((df3.isnull().sum()*100)/len(df3),2).astype(str)+"5")
missingdf3

Missing Values Percentage of missing values
given_name                 0                         0.05
surname                    0                         0.05
auralin                    0                         0.05
novodra                    0                         0.05
hba1c_start                0                         0.05
hba1c_end                  0                         0.05
hba1c_change              28                        40.05

In [7]:
df4=pd.read_csv(r"D:\Python\Datasets\adverse_reaction.csv")
df4.head()

given_name     surname           adverse_reaction
0      berta  napolitani  injection site discomfort
1       lena        baer               hypoglycemia
2     joseph         day               hypoglycemia
3     flavia  fiorentino                      cough
4    manouck     wubbels          throat irritation

In [55]:
missingdf4=df4.isnull().sum().to_frame().rename(columns={0:"Missing Values"})
missingdf4["Percentage of missing values"]=(round((df4.isnull().sum()*100)/len(df4),2).astype(str)+"%")
missingdf4

Missing Values Percentage of missing values
given_name                     0                         0.0%
surname                        0                         0.0%
adverse_reaction               0                         0.0%

# Write a summary for your data

This is a dataset about 500 patients of which 350 patients participated in a clinical trial. None of the patients were using Novodra(a popular injectable insulin) or Auralin (the oral insulin being researched) as their primary source of insulin before. All were expecting elevated HbA1c levels.<br>

All 350 patients were treated with Novodra to establish a baseline HbA1c level and insulin dose. After 4 weeks, which isn't enough time to capture all the change in HbA1c that can be attributed by the switch to Auralin or Novodra:

* 175 patients switched to Auralin for 24 weeks.
* 175 patients continued using Novodra for 24 weeks.

Data about patients feeling some adverse effects is also recorded.





# Write Column Descriptions

1. **Assigned Sex (`assigned_sex`):**
   - Represents the assigned sex of the patient.
   - Examples: Female, Male.

2. **Given Name (`given_name`):**
   - Indicates the given or first name of the patient.
   - Examples: Zoe, Pamela, Jae, Liêm, Tim.

3. **Surname (`surname`):**
   - Represents the surname or last name of the patient.
   - Examples: Wellish, Hill, Debord, Phan, Neudorf.

4. **Address (`address`):**
   - Specifies the street address of the patient.
   - Examples: 576 Brown Bear Drive, 2370 University Hill Road, 1493 Poling Farm Road, 2335 Webster Street, 1428 Turkey Pen Lane.

5. **City (`city`):**
   - Indicates the city of residence for the patient.
   - Examples: Rancho California, Armstrong, York, Woodbridge, Dothan.

6. **State (`state`):**
   - Specifies the state or region of the patient's residence.
   - Examples: California, Illinois, Nebraska, NJ, AL.

7. **ZIP Code (`zip_code`):**
   - Represents the ZIP code or postal code of the patient's address.
   - Examples: 92390.0, 61812.0, 68467.0, 7095.0, 36303.0.

8. **Country (`country`):**
   - Specifies the country where the patient resides.
   - Example: United States.

9. **Contact Information (`contact`):**
   - Represents the contact information (email and phone) of the patient.
   - Examples: ZoeWellish@superrito.com, PamelaSHill@cuvox.de+1 (217) 569-3204, JaeMDebord@gustr.com, PhanBaLiem@jourrapide.com+1 (732) 636-8246, TimNeudorf@cuvox.de.

10. **Birthdate (`birthdate`):**
    - Represents the birthdate of the patient.
    - Examples: 7/10/1976, 4/3/1967, 2/19/1980, 7/26/1951, 2/18/1928.

11. **Weight (`weight`):**
    - Represents the weight of the patient.
    - Examples: 121.7, 118.8, 177.8, 220.9, 192.3.

12. **Height (`height`):**
    - Indicates the height of the patient.
    - Examples: 66, 66, 71, 70, 27.

13. **BMI (`bmi`):**
    - Represents the Body Mass Index (BMI) of the patient.
    - Examples: 19.6, 19.2, 24.8, 31.7, 26.1.

This format aims to provide a clear and concise overview of each column in a structured manner.

# Add any additional information
* Insulin resistance varies person to person, which is why both starting median daily dose and ending median daily dose are required, i.e., to calculate the change in the dosage.
* It is important to test drugs and medical products in the people they are meant to help. People of different age, race, sex, and ethnic group must be included in clinical trials.This density is reflected int the patients table. 

# Types of Assessment.

* Manual-Looking through the data manually in google sheets.
* Programmatic-By using pandas functions such as info(), describe() or sample().

# Steps in Assessment.
There are 2 steps involved in Assessment
* Discover
* Document


In [8]:
# Export data for manual assessment
with pd.ExcelWriter("clinical_trials.xlsx") as writer:
    df1.to_excel(writer,sheet_name="patients")
    df2.to_excel(writer,sheet_name="treatments")
    df3.to_excel(writer,sheet_name="treatment_cut")
    df4.to_excel(writer,sheet_name="adverse_reaction")

# Issues with data
# 1) Dirty Data

<h4>Table ->df1</h4><ol>
<li>patient_id 9 has misspelled name "David" as "Dsvid"-Accuracy issue.</li>
<li>state column sometimes contain full name sometime abbreviations-Consistency issue</li>
<li>zip code column has entries with 4 digit- validity issue.</li>
<li>Data missing in address, city, state, zip code, country, contact columns- Completion issue</li>
<li>Incorrect data types assigned to sex, zip code and DOB-validity issue</li>
<li>Duplicated entries in the names of John Doe-Accuracy issue.</li>
<li>One patient has weight 48 pounds-Accuracy issue</li>
<li>One patient has height 27 inches-Accuracy issue</li>
</ol>
<h4>Table ->df2 and df3</h4><ol>
<li>Name and surname columns are in lower case-Consistency issue</li>
<li>hyphen is treated as Nan-validity issue</li>
<li>Remove u from auralin and novodra column-validity issue</li>
<li>Missing values in hba1c column-Completion issue</li>
<li>Duplicated entries in the names of joseph day-Accuracy issue.</li>
<li>In hbaic_changes 9 is written instead of 4-Accuracy issue.</li>
</ol>

<h4>Table ->df4</h4><ol>
<li>Name and surname columns are in lower case-Consistency issue</li>
</ol>
</p>

# Messy data

<p><h4>Table ->df1</h4><ol>
<li>contact data contains both email and phone number.</li></ol></p>
<li>Merge both the data.</li>
<h4>Table ->df2 and df3</h4><ol>
<li>auralin and novodra column should split into two columns start and end dose</li></ol>
<h4>Table ->df4</h4><ol>
<li>This table should not exist independently</li>
</ol>
</p>

# Automatic Assessment
<ul>
<li> head and tail</li>
<li>sample</li>
<li>info</li>
<li>isnull</li>
<li>duplicated</li>
<li>describe</li>
</ul>

In [9]:
# Gives both head(5) and tail(5)
df1


assigned_sex given_name      surname                    address  \
patient_id                                                                   
1                female        Zoe      Wellish       576 Brown Bear Drive   
2                female     Pamela         Hill  2370 University Hill Road   
3                  male        Jae       Debord      1493 Poling Farm Road   
4                  male       Liêm         Phan        2335 Webster Street   
5                  male        Tim      Neudorf       1428 Turkey Pen Lane   
...                 ...        ...          ...                        ...   
499                male    Mustafa    Lindström        2530 Victoria Court   
500                male      Ruman      Bisliev   494 Clarksburg Park Road   
501              female      Jinke    de Keizer          649 Nutter Street   
502              female    Chidalu  Onyekaozulu   3652 Boone Crockett Lane   
503                male        Pat      Gersten          2778 North Avenue   

                         city       state  zip_code        country  \
patient_id                                                           
1           Rancho California  California   92390.0  United States   
2                   Armstrong    Illinois   61812.0  United States   
3                        York    Nebraska   68467.0  United States   
4                  Woodbridge          NJ    7095.0  United States   
5                      Dothan          AL   36303.0  United States   
...                       ...         ...       ...            ...   
499              Milton Mills          ME    3852.0  United States   
500                    Sedona          AZ   86341.0  United States   
501             Overland Park          MO   64110.0  United States   
502                   Seattle          WA   98109.0  United States   
503                      Burr    Nebraska   68324.0  United States   

                                                    contact  birthdate  \
patient_id                                                               
1                      951-719-9170ZoeWellish@superrito.com  7/10/1976   
2                     PamelaSHill@cuvox.de+1 (217) 569-3204   4/3/1967   
3                          402-363-6804JaeMDebord@gustr.com  2/19/1980   
4                PhanBaLiem@jourrapide.com+1 (732) 636-8246  7/26/1951   
5                           334-515-7487TimNeudorf@cuvox.de  2/18/1928   
...                                                     ...        ...   
499             207-477-0579MustafaLindstrom@jourrapide.com  4/10/1959   
500                      928-284-4492RumanBisliev@gustr.com  3/26/1948   
501                   816-223-6007JinkedeKeizer@teleworm.us  1/13/1971   
502         ChidaluOnyekaozulu@jourrapide.com1 360 443 2060  2/13/1952   
503                    PatrickGersten@rhyta.com402-848-4923   5/3/1954   

            weight  height   bmi  
patient_id                        
1            121.7      66  19.6  
2            118.8      66  19.2  
3            177.8      71  24.8  
4            220.9      70  31.7  
5            192.3      27  26.1  
...            ...     ...   ...  
499          181.1      72  24.6  
500          239.6      70  34.4  
501          171.2      67  26.8  
502          176.9      67  27.7  
503          138.2      71  19.3  

[503 rows x 13 columns]

In [10]:
df1.sample(10)

assigned_sex    given_name      surname                  address  \
patient_id                                                                    
355              female        Vivian        House       4932 Goldleaf Lane   
292                male          Enco       Žibrik         4792 Maud Street   
463                male         Guðni    Heimisson  430 Canis Heights Drive   
54                 male  Kwemtochukwu    Ogochukwu         2172 Lynn Street   
36               female        Kamila     Pecinová     3558 Longview Avenue   
503                male           Pat      Gersten        2778 North Avenue   
348              female         Urška     Aličajić       3781 Hamill Avenue   
385                male          Even      Knutsen         4851 Andy Street   
172              female       Blagica     Klobučar      3920 Braxton Street   
466                male        Gabryŝ  Tomaszewski        1343 Clair Street   

                      city       state  zip_code        country  \
patient_id                                                        
355                 Newark          NJ    7102.0  United States   
292         Camden Wyoming          DE   19934.0  United States   
463            Los Angeles  California   90007.0  United States   
54                Franklin          MA    2038.0  United States   
36                New York    New York   10004.0  United States   
503                   Burr    Nebraska   68324.0  United States   
348              San Diego          CA   92109.0  United States   
385                 Custer          SD   57730.0  United States   
172              La Grange          IL   60525.0  United States   
466              Hillsboro          TX   76645.0  United States   

                                                 contact   birthdate  weight  \
patient_id                                                                     
355                  201-586-2848VivianRHouse@dayrep.com   8/13/1936   130.2   
292                 302-698-2057EncoZibrik@superrito.com   1/25/1983   203.9   
463              213-749-6958GuniHeimisson@superrito.com  11/28/1983   154.9   
54          617-317-5055KwemtochukwuOgochukwu@einrot.com   6/30/1976   150.5   
36                 718-501-0503KamilaPecinova@dayrep.com  12/23/1985   198.9   
503                 PatrickGersten@rhyta.com402-848-4923    5/3/1954   138.2   
348                  858-273-2043UrskaAlicajic@rhyta.com   6/27/1952   183.3   
385               EvenKnutsen@rhyta.com+1 (605) 440-5492  10/26/1972   180.2   
172               815-533-7692BlagicaKlobucar@einrot.com   11/6/1921   106.5   
466            GabrysTomaszewski@armyspy.com254-582-1795   9/16/1998   160.8   

            height   bmi  
patient_id                
355             62  23.8  
292             74  26.2  
463             71  21.6  
54              72  20.4  
36              62  36.4  
503             71  19.3  
348             63  32.5  
385             74  23.1  
172             59  21.5  
466             67  25.2

In [11]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
Index: 503 entries, 1 to 503
Data columns (total 13 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   assigned_sex  503 non-null    object 
 1   given_name    503 non-null    object 
 2   surname       503 non-null    object 
 3   address       491 non-null    object 
 4   city          491 non-null    object 
 5   state         491 non-null    object 
 6   zip_code      491 non-null    float64
 7   country       491 non-null    object 
 8   contact       491 non-null    object 
 9   birthdate     503 non-null    object 
 10  weight        503 non-null    float64
 11  height        503 non-null    int64  
 12  bmi           503 non-null    float64
dtypes: float64(3), int64(1), object(9)
memory usage: 55.0+ KB


In [12]:
df1[df1["address"].isnull()]

assigned_sex given_name      surname address city state  zip_code  \
patient_id                                                                     
210              female     Lalita  Eldarkhanov     NaN  NaN   NaN       NaN   
220                male         Mỹ        Quynh     NaN  NaN   NaN       NaN   
231              female  Elisabeth      Knudsen     NaN  NaN   NaN       NaN   
235              female    Martina    Tománková     NaN  NaN   NaN       NaN   
243                male       John      O'Brian     NaN  NaN   NaN       NaN   
250                male   Benjamin       Mehler     NaN  NaN   NaN       NaN   
258                male        Jin         Kung     NaN  NaN   NaN       NaN   
265              female   Wafiyyah       Asfour     NaN  NaN   NaN       NaN   
270              female     Flavia   Fiorentino     NaN  NaN   NaN       NaN   
279              female   Generosa        Cabán     NaN  NaN   NaN       NaN   
287                male      Lewis         Webb     NaN  NaN   NaN       NaN   
297              female        Chỉ          Lâm     NaN  NaN   NaN       NaN   

           country contact   birthdate  weight  height   bmi  
patient_id                                                    
210            NaN     NaN   8/14/1950   143.4      62  26.2  
220            NaN     NaN    4/9/1978   237.8      69  35.1  
231            NaN     NaN   9/23/1976   165.9      63  29.4  
235            NaN     NaN    4/7/1936   199.5      65  33.2  
243            NaN     NaN   2/25/1957   205.3      74  26.4  
250            NaN     NaN  10/30/1951   146.5      69  21.6  
258            NaN     NaN   5/17/1995   231.7      69  34.2  
265            NaN     NaN   11/3/1989   158.6      63  28.1  
270            NaN     NaN   10/9/1937   175.2      61  33.1  
279            NaN     NaN  12/16/1962   124.3      69  18.4  
287            NaN     NaN    4/1/1979   155.3      68  23.6  
297            NaN     NaN   5/14/1990   181.1      63  32.1

In [13]:
df4.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 34 entries, 0 to 33
Data columns (total 3 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   given_name        34 non-null     object
 1   surname           34 non-null     object
 2   adverse_reaction  34 non-null     object
dtypes: object(3)
memory usage: 948.0+ bytes


In [14]:
df1[df1.duplicated(subset=["given_name","surname"])]

assigned_sex given_name surname          address      city state  \
patient_id                                                                    
230                male       John     Doe  123 Main Street  New York    NY   
238                male       John     Doe  123 Main Street  New York    NY   
245                male       John     Doe  123 Main Street  New York    NY   
252                male       John     Doe  123 Main Street  New York    NY   
278                male       John     Doe  123 Main Street  New York    NY   

            zip_code        country                      contact birthdate  \
patient_id                                                                   
230          12345.0  United States  johndoe@email.com1234567890  1/1/1975   
238          12345.0  United States  johndoe@email.com1234567890  1/1/1975   
245          12345.0  United States  johndoe@email.com1234567890  1/1/1975   
252          12345.0  United States  johndoe@email.com1234567890  1/1/1975   
278          12345.0  United States  johndoe@email.com1234567890  1/1/1975   

            weight  height   bmi  
patient_id                        
230          180.0      72  24.4  
238          180.0      72  24.4  
245          180.0      72  24.4  
252          180.0      72  24.4  
278          180.0      72  24.4

In [15]:
df2[df2.duplicated(subset=["given_name","surname"])]

given_name surname    auralin novodra  hba1c_start  hba1c_end  \
136     joseph     day  29u - 36u       -          7.7       7.19   

     hba1c_change  
136           NaN

In [16]:
df3.duplicated(subset=["given_name","surname"]).sum()

0

In [17]:
df4.duplicated(subset=["given_name","surname"]).sum()

0

In [18]:
df1.describe()

zip_code      weight      height         bmi
count    491.000000  503.000000  503.000000  503.000000
mean   49084.118126  173.434990   66.634195   27.483897
std    30265.807442   33.916741    4.411297    5.276438
min     1002.000000   48.800000   27.000000   17.100000
25%    21920.500000  149.300000   63.000000   23.300000
50%    48057.000000  175.300000   67.000000   27.200000
75%    75679.000000  199.500000   70.000000   31.750000
max    99701.000000  255.900000   79.000000   37.700000

In [19]:
df1[df1["weight"]==48.800000]

assigned_sex given_name   surname              address     city  \
patient_id                                                                   
211              female    Camilla  Zaitseva  4689 Briarhill Lane  Wooster   

           state  zip_code        country  \
patient_id                                  
211           OH   44691.0  United States   

                                              contact   birthdate  weight  \
patient_id                                                                  
211         330-202-2145CamillaZaitseva@superrito.com  11/26/1938    48.8   

            height   bmi  
patient_id                
211             63  19.1

In [20]:
df1[df1["height"]==27]

assigned_sex given_name  surname               address    city  \
patient_id                                                                  
5                  male        Tim  Neudorf  1428 Turkey Pen Lane  Dothan   

           state  zip_code        country                          contact  \
patient_id                                                                   
5             AL   36303.0  United States  334-515-7487TimNeudorf@cuvox.de   

            birthdate  weight  height   bmi  
patient_id                                   
5           2/18/1928   192.3      27  26.1

In [21]:
df2.describe()

hba1c_start   hba1c_end  hba1c_change
count   280.000000  280.000000    171.000000
mean      7.985929    7.589286      0.546023
std       0.568638    0.569672      0.279555
min       7.500000    7.010000      0.200000
25%       7.660000    7.270000      0.340000
50%       7.800000    7.420000      0.380000
75%       7.970000    7.570000      0.920000
max       9.950000    9.580000      0.990000

In [22]:
df2.sort_values("hba1c_start")

given_name      surname    auralin    novodra  hba1c_start  hba1c_end  \
270       mika   martinsson  34u - 43u          -         7.50       7.17   
113       kari  laatikainen  39u - 43u          -         7.50       7.11   
126     jowita   wiśniewska          -  22u - 23u         7.50       7.08   
53      nasser      mansour          -  33u - 31u         7.51       7.06   
105     finlay     sheppard          -  31u - 30u         7.51       7.17   
..         ...          ...        ...        ...          ...        ...   
25      benoît       bonami          -  44u - 43u         9.82       9.40   
171    justyna    kowalczyk  24u - 34u          -         9.84       9.44   
81      robert       wagner  43u - 49u          -         9.84       9.52   
75   mackenzie        mckay          -  44u - 43u         9.87       9.48   
166      annie        allen  36u - 42u          -         9.95       9.58   

     hba1c_change  
270          0.33  
113           NaN  
126          0.92  
53           0.95  
105          0.34  
..            ...  
25           0.92  
171           NaN  
81           0.32  
75           0.39  
166          0.37  

[280 rows x 7 columns]

In [23]:
df2.sort_values("hba1c_end")

given_name    surname    auralin    novodra  hba1c_start  hba1c_end  \
39   krisztina     magyar          -  32u - 27u         7.52       7.01   
80      hideki  haraguchi          -  37u - 35u         7.59       7.05   
187       león    reynoso          -  38u - 40u         7.59       7.06   
53      nasser    mansour          -  33u - 31u         7.51       7.06   
86     ananías   enríquez          -  44u - 45u         7.58       7.07   
..         ...        ...        ...        ...          ...        ...   
192      valur  bjarkason          -  31u - 36u         9.71       9.41   
171    justyna  kowalczyk  24u - 34u          -         9.84       9.44   
75   mackenzie      mckay          -  44u - 43u         9.87       9.48   
81      robert     wagner  43u - 49u          -         9.84       9.52   
166      annie      allen  36u - 42u          -         9.95       9.58   

     hba1c_change  
39           0.51  
80           0.54  
187          0.53  
53           0.95  
86           0.51  
..            ...  
192          0.30  
171           NaN  
75           0.39  
81           0.32  
166          0.37  

[280 rows x 7 columns]

In [24]:
df2.sort_values("hba1c_change",na_position="first")

given_name      surname    auralin    novodra  hba1c_start  hba1c_end  \
0     veronika     jindrová  41u - 48u          -         7.63       7.20   
2     yukitaka     takenaka          -  39u - 36u         7.68       7.25   
8        saber       ménard          -  54u - 54u         8.08       7.70   
9         asia      woźniak  30u - 36u          -         7.76       7.37   
10      joseph          day  29u - 36u          -         7.70       7.19   
..         ...          ...        ...        ...          ...        ...   
49     jackson      addison          -  42u - 42u         7.99       7.51   
17        gina         cain          -  36u - 36u         7.88       7.40   
32       laura  ehrlichmann          -  43u - 40u         7.95       7.46   
245         wu         sung          -  47u - 48u         7.61       7.12   
138    giovana        rocha          -  23u - 21u         7.87       7.38   

     hba1c_change  
0             NaN  
2             NaN  
8             NaN  
9             NaN  
10            NaN  
..            ...  
49           0.98  
17           0.98  
32           0.99  
245          0.99  
138          0.99  

[280 rows x 7 columns]

# Assessing Data is an iterative process.

# Data Quality Dimensions
<p><h5>Completeness:Consistency issue</h5>
<h5>Validity:Is data invalid ->negative height->duplicate patient id.</h5>
<h5>Accuracy:Data is valid but not accurate->1kg.</h5>
<h5>Consistency:Both valid and accurate but written differently->New York and NY</h5></p>

# Order of Severity
Completeness->Validity->Accuracy->Consistency

# Data Cleaning Order
<p><ol>
<li>Quality->Completeness</li>
<li>Tidiness</li>
<li>Quality->Validity</li>
<li>Quality->Accuracy</li>
<li>Quality->Consistency</li>
</ol></p>

# Steps involved in Data Cleaning.

* Defining the problem.
* Coding the problem.
* Testing the problem.

Always make sure to create a copy of your pandas dataframe before you start the cleaning process.


In [58]:
df1_copy=df1.copy()
df2_copy=df2.copy()
df3_copy=df3.copy()
df4_copy=df4.copy()

# Define the problem
-> Replace all missing values of df1 with no data.
-> Sub hba1c_start from hba1c_end to get all the change values

In [64]:
# code
df1_copy.fillna("No data",inplace=True)

0

In [65]:
# Test
df1_copy.info()

<class 'pandas.core.frame.DataFrame'>
Index: 503 entries, 1 to 503
Data columns (total 13 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   assigned_sex  503 non-null    object 
 1   given_name    503 non-null    object 
 2   surname       503 non-null    object 
 3   address       503 non-null    object 
 4   city          503 non-null    object 
 5   state         503 non-null    object 
 6   zip_code      503 non-null    object 
 7   country       503 non-null    object 
 8   contact       503 non-null    object 
 9   birthdate     503 non-null    object 
 10  weight        503 non-null    float64
 11  height        503 non-null    int64  
 12  bmi           503 non-null    float64
dtypes: float64(2), int64(1), object(10)
memory usage: 55.0+ KB


In [66]:
df2.head()

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
0   veronika    jindrová  41u - 48u          -         7.63       7.20   
1     elliot  richardson          -  40u - 45u         7.56       7.09   
2   yukitaka    takenaka          -  39u - 36u         7.68       7.25   
3       skye  gormanston  33u - 36u          -         7.97       7.62   
4     alissa      montez          -  33u - 29u         7.78       7.46   

   hba1c_change  Percentage of Missing values  
0           NaN                           NaN  
1          0.97                           NaN  
2           NaN                           NaN  
3          0.35                           NaN  
4          0.32                           NaN

In [74]:
# code
df2_copy["hba1c_change"]=df2_copy["hba1c_start"]-df2_copy["hba1c_end"]
# df2_copy.drop("Percentage of Missing values",axis=1,inplace=True)

In [77]:
df2_copy.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 280 entries, 0 to 279
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   given_name    280 non-null    object 
 1   surname       280 non-null    object 
 2   auralin       280 non-null    object 
 3   novodra       280 non-null    object 
 4   hba1c_start   280 non-null    float64
 5   hba1c_end     280 non-null    float64
 6   hba1c_change  280 non-null    float64
dtypes: float64(3), object(4)
memory usage: 15.4+ KB


In [76]:
df3_copy["hba1c_change"]=df3_copy["hba1c_start"]-df3_copy["hba1c_end"]
df3_copy.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 70 entries, 0 to 69
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   given_name    70 non-null     object 
 1   surname       70 non-null     object 
 2   auralin       70 non-null     object 
 3   novodra       70 non-null     object 
 4   hba1c_start   70 non-null     float64
 5   hba1c_end     70 non-null     float64
 6   hba1c_change  70 non-null     float64
dtypes: float64(3), object(4)
memory usage: 4.0+ KB


In [80]:
df1_copy

assigned_sex given_name      surname                    address  \
patient_id                                                                   
1                female        Zoe      Wellish       576 Brown Bear Drive   
2                female     Pamela         Hill  2370 University Hill Road   
3                  male        Jae       Debord      1493 Poling Farm Road   
4                  male       Liêm         Phan        2335 Webster Street   
5                  male        Tim      Neudorf       1428 Turkey Pen Lane   
...                 ...        ...          ...                        ...   
499                male    Mustafa    Lindström        2530 Victoria Court   
500                male      Ruman      Bisliev   494 Clarksburg Park Road   
501              female      Jinke    de Keizer          649 Nutter Street   
502              female    Chidalu  Onyekaozulu   3652 Boone Crockett Lane   
503                male        Pat      Gersten          2778 North Avenue   

                         city       state zip_code        country  \
patient_id                                                          
1           Rancho California  California  92390.0  United States   
2                   Armstrong    Illinois  61812.0  United States   
3                        York    Nebraska  68467.0  United States   
4                  Woodbridge          NJ   7095.0  United States   
5                      Dothan          AL  36303.0  United States   
...                       ...         ...      ...            ...   
499              Milton Mills          ME   3852.0  United States   
500                    Sedona          AZ  86341.0  United States   
501             Overland Park          MO  64110.0  United States   
502                   Seattle          WA  98109.0  United States   
503                      Burr    Nebraska  68324.0  United States   

                                                    contact  birthdate  \
patient_id                                                               
1                      951-719-9170ZoeWellish@superrito.com  7/10/1976   
2                     PamelaSHill@cuvox.de+1 (217) 569-3204   4/3/1967   
3                          402-363-6804JaeMDebord@gustr.com  2/19/1980   
4                PhanBaLiem@jourrapide.com+1 (732) 636-8246  7/26/1951   
5                           334-515-7487TimNeudorf@cuvox.de  2/18/1928   
...                                                     ...        ...   
499             207-477-0579MustafaLindstrom@jourrapide.com  4/10/1959   
500                      928-284-4492RumanBisliev@gustr.com  3/26/1948   
501                   816-223-6007JinkedeKeizer@teleworm.us  1/13/1971   
502         ChidaluOnyekaozulu@jourrapide.com1 360 443 2060  2/13/1952   
503                    PatrickGersten@rhyta.com402-848-4923   5/3/1954   

            weight  height   bmi  
patient_id                        
1            121.7      66  19.6  
2            118.8      66  19.2  
3            177.8      71  24.8  
4            220.9      70  31.7  
5            192.3      27  26.1  
...            ...     ...   ...  
499          181.1      72  24.6  
500          239.6      70  34.4  
501          171.2      67  26.8  
502          176.9      67  27.7  
503          138.2      71  19.3  

[503 rows x 13 columns]

Merge df2 and df3

In [96]:
treatment_df=pd.concat([df2,df3])
treatment_df.drop("Percentage of Missing values",axis=1,inplace=True)

In [97]:
treatment_df

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
0    veronika    jindrová  41u - 48u          -         7.63       7.20   
1      elliot  richardson          -  40u - 45u         7.56       7.09   
2    yukitaka    takenaka          -  39u - 36u         7.68       7.25   
3        skye  gormanston  33u - 36u          -         7.97       7.62   
4      alissa      montez          -  33u - 29u         7.78       7.46   
..        ...         ...        ...        ...          ...        ...   
65     rovzan     kishiev  32u - 37u          -         7.75       7.41   
66      jakob    jakobsen          -  28u - 26u         7.96       7.51   
67      bernd   schneider  48u - 56u          -         7.74       7.44   
68      berta  napolitani          -  42u - 44u         7.68       7.21   
69     armina       sauvé  36u - 46u          -         7.86       7.40   

    hba1c_change  
0            NaN  
1           0.97  
2            NaN  
3           0.35  
4           0.32  
..           ...  
65          0.34  
66          0.95  
67          0.30  
68           NaN  
69           NaN  

[350 rows x 7 columns]

In [106]:
#treatment_df=treatment_df.melt(id_vars=["given_name","surname","hba1c_start","hba1c_end","hba1c_change"],var_name="type",value_name="Dosage_range")


In [110]:
treatment_df=treatment_df[treatment_df["Dosage_range"]!="-"]
treatment_df

given_name     surname  hba1c_start  hba1c_end  hba1c_change     type  \
0       veronika    jindrová         7.63       7.20           NaN  auralin   
3           skye  gormanston         7.97       7.62          0.35  auralin   
6         sophia      haugen         7.65       7.27          0.38  auralin   
7          eddie      archer         7.89       7.55          0.34  auralin   
9           asia     woźniak         7.76       7.37           NaN  auralin   
..           ...         ...          ...        ...           ...      ...   
688  christopher    woodward         7.51       7.06          0.95  novodra   
690        maret    sultygov         7.67       7.30           NaN  novodra   
694        lixue       hsueh         9.21       8.80          0.91  novodra   
696        jakob    jakobsen         7.96       7.51          0.95  novodra   
698        berta  napolitani         7.68       7.21           NaN  novodra   

    Dosage_range  
0      41u - 48u  
3      33u - 36u  
6      37u - 42u  
7      31u - 38u  
9      30u - 36u  
..           ...  
688    55u - 51u  
690    26u - 23u  
694    22u - 23u  
696    28u - 26u  
698    42u - 44u  

[350 rows x 7 columns]

In [114]:
treatment_df["Dosage_start"]=treatment_df["Dosage_range"].str.split("-").str.get(0)
treatment_df["Dosage_end"]=treatment_df["Dosage_range"].str.split("-").str.get(1)

In [116]:
treatment_df.drop("Dosage_range",axis=1)

given_name     surname  hba1c_start  hba1c_end  hba1c_change     type  \
0       veronika    jindrová         7.63       7.20           NaN  auralin   
3           skye  gormanston         7.97       7.62          0.35  auralin   
6         sophia      haugen         7.65       7.27          0.38  auralin   
7          eddie      archer         7.89       7.55          0.34  auralin   
9           asia     woźniak         7.76       7.37           NaN  auralin   
..           ...         ...          ...        ...           ...      ...   
688  christopher    woodward         7.51       7.06          0.95  novodra   
690        maret    sultygov         7.67       7.30           NaN  novodra   
694        lixue       hsueh         9.21       8.80          0.91  novodra   
696        jakob    jakobsen         7.96       7.51          0.95  novodra   
698        berta  napolitani         7.68       7.21           NaN  novodra   

    Dosage_start Dosage_end  
0           41u         48u  
3           33u         36u  
6           37u         42u  
7           31u         38u  
9           30u         36u  
..           ...        ...  
688         55u         51u  
690         26u         23u  
694         22u         23u  
696         28u         26u  
698         42u         44u  

[350 rows x 8 columns]

In [119]:
treatment_df["Dosage_start"]=treatment_df["Dosage_start"].str.replace("u","")
treatment_df["Dosage_end"]=treatment_df["Dosage_end"].str.replace("u","")

In [122]:
treatment_df["Dosage_start"]=treatment_df["Dosage_start"].astype(int)
treatment_df["Dosage_end"]=treatment_df["Dosage_end"].astype(int)

In [124]:
treatment_df.drop("Dosage_range",axis=1,inplace=True)

In [125]:
treatment_df

given_name     surname  hba1c_start  hba1c_end  hba1c_change     type  \
0       veronika    jindrová         7.63       7.20           NaN  auralin   
3           skye  gormanston         7.97       7.62          0.35  auralin   
6         sophia      haugen         7.65       7.27          0.38  auralin   
7          eddie      archer         7.89       7.55          0.34  auralin   
9           asia     woźniak         7.76       7.37           NaN  auralin   
..           ...         ...          ...        ...           ...      ...   
688  christopher    woodward         7.51       7.06          0.95  novodra   
690        maret    sultygov         7.67       7.30           NaN  novodra   
694        lixue       hsueh         9.21       8.80          0.91  novodra   
696        jakob    jakobsen         7.96       7.51          0.95  novodra   
698        berta  napolitani         7.68       7.21           NaN  novodra   

     Dosage_start  Dosage_end  
0              41          48  
3              33          36  
6              37          42  
7              31          38  
9              30          36  
..            ...         ...  
688            55          51  
690            26          23  
694            22          23  
696            28          26  
698            42          44  

[350 rows x 8 columns]

In [130]:
treatment_df=treatment_df.merge(df4,how="left",on=["given_name","surname"])
treatment_df

given_name     surname  hba1c_start  hba1c_end  hba1c_change     type  \
0       veronika    jindrová         7.63       7.20           NaN  auralin   
1           skye  gormanston         7.97       7.62          0.35  auralin   
2         sophia      haugen         7.65       7.27          0.38  auralin   
3          eddie      archer         7.89       7.55          0.34  auralin   
4           asia     woźniak         7.76       7.37           NaN  auralin   
..           ...         ...          ...        ...           ...      ...   
345  christopher    woodward         7.51       7.06          0.95  novodra   
346        maret    sultygov         7.67       7.30           NaN  novodra   
347        lixue       hsueh         9.21       8.80          0.91  novodra   
348        jakob    jakobsen         7.96       7.51          0.95  novodra   
349        berta  napolitani         7.68       7.21           NaN  novodra   

     Dosage_start  Dosage_end         adverse_reaction_x  \
0              41          48                        NaN   
1              33          36                        NaN   
2              37          42                        NaN   
3              31          38                        NaN   
4              30          36                        NaN   
..            ...         ...                        ...   
345            55          51                     nausea   
346            26          23                        NaN   
347            22          23  injection site discomfort   
348            28          26               hypoglycemia   
349            42          44  injection site discomfort   

            adverse_reaction_y  
0                          NaN  
1                          NaN  
2                          NaN  
3                          NaN  
4                          NaN  
..                         ...  
345                     nausea  
346                        NaN  
347  injection site discomfort  
348               hypoglycemia  
349  injection site discomfort  

[350 rows x 10 columns]

In [128]:
df4

given_name     surname           adverse_reaction
0         berta  napolitani  injection site discomfort
1          lena        baer               hypoglycemia
2        joseph         day               hypoglycemia
3        flavia  fiorentino                      cough
4       manouck     wubbels          throat irritation
5       jasmine       sykes               hypoglycemia
6        louise     johnson               hypoglycemia
7       albinca     komavec               hypoglycemia
8           noe      aranda               hypoglycemia
9         sofia   hermansen  injection site discomfort
10        tegan     johnson                   headache
11         abel     yonatan                      cough
12    abdul-nur         isa               hypoglycemia
13         leon      scholz  injection site discomfort
14     gabriele     saenger               hypoglycemia
15       jia li        teng                     nausea
16        jakob    jakobsen               hypoglycemia
17  christopher    woodward                     nausea
18          ole    petersen               hypoglycemia
19       finley    chandler                   headache
20     anenechi       chidi               hypoglycemia
21     miłosław  wiśniewski  injection site discomfort
22        lixue       hsueh  injection site discomfort
23        merci      leroux               hypoglycemia
24         kang         mai  injection site discomfort
25       elliot  richardson               hypoglycemia
26      clinton      miller          throat irritation
27       idalia       moore               hypoglycemia
28       xiuxiu       chang               hypoglycemia
29         alex    crawford               hypoglycemia
30       monika      lončar               hypoglycemia
31       steven         roy                   headache
32      cecilie      nilsen               hypoglycemia
33    krisztina      magyar               hypoglycemia